// @pooltogether/pooltogether-proxy-factory-package/dist/src/deploy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.factoryDeploy = void 0;
// exports a generic factoryDeploy using the generic minimal proxy factories
const chalk_1 = __importDefault(require("chalk"));
const fs_1 = require("fs");
const GenericProxyFactory_json_1 = __importDefault(require("../abis/GenericProxyFactory.json"));
const getGenericProxyFactoryForNetwork_1 = require("./helpers/getGenericProxyFactoryForNetwork");
const ethers_1 = require("ethers");
const displayLogs = !process.env.HIDE_DEPLOY_LOG;
function dim(logMessage) {
    if (displayLogs) {
        console.log(chalk_1.default.dim(logMessage));
    }
}
function cyan(logMessage) {
    if (displayLogs) {
        console.log(chalk_1.default.cyan(logMessage));
    }
}
function yellow(logMessage) {
    if (displayLogs) {
        console.log(chalk_1.default.yellow(logMessage));
    }
}
function green(logMessage) {
    if (displayLogs) {
        console.log(chalk_1.default.green(logMessage));
    }
}
function factoryDeploy(deploySettings) {
    return __awaiter(this, void 0, void 0, function* () {
        const provider = deploySettings.provider;
        let networkName = (yield provider.getNetwork()).name;
        let initializeData;
        if (!deploySettings.initializeData) {
            initializeData = "0x";
        }
        else {
            initializeData = deploySettings.initializeData;
        }
        let skipIfAlreadyDeployed;
        if (deploySettings.skipIfAlreadyDeployed === undefined) {
            skipIfAlreadyDeployed = false;
        }
        else {
            skipIfAlreadyDeployed = true;
        }
        if (skipIfAlreadyDeployed && fs_1.existsSync(`./deployments/${networkName}/${deploySettings.contractName}.json`)) {
            cyan(`Using existing implementation for ${deploySettings.contractName}`);
            // contract already exists -- read info for return object
            const previouslyDeployedContract = JSON.parse(yield fs_1.readFileSync(`./deployments/${networkName}/${deploySettings.contractName}.json`, { encoding: "utf-8" }));
            return {
                newlyDeployed: false,
                address: previouslyDeployedContract.address,
                transactionHash: previouslyDeployedContract.transactionHash,
                receipt: previouslyDeployedContract.receipt,
            };
        }
        // get address of minimal proxy factory
        const chainId = (yield provider.getNetwork()).chainId;
        if (chainId === 31337 || chainId === 1337) { // network name "unknown" for the test networks
            networkName = "localhost";
        }
        let genericProxyFactoryAddress = getGenericProxyFactoryForNetwork_1.getGenericProxyFactoryAddressForChainId(chainId);
        // if no generic proxy factory -- create one
        if (genericProxyFactoryAddress == "") {
            dim(`No GenericProxyFactory deployment found. Deploying a new GenericProxyFactory`);
            const genericProxyFactoryInterface = new ethers_1.ethers.utils.Interface(GenericProxyFactory_json_1.default.abi);
            const genericProxyFactoryContractFactory = new ethers_1.ethers.ContractFactory(genericProxyFactoryInterface, GenericProxyFactory_json_1.default.bytecode, deploySettings.signer);
            const genericProxyFacoryContract = yield genericProxyFactoryContractFactory.deploy();
            yield genericProxyFacoryContract.deployTransaction.wait();
            genericProxyFactoryAddress = genericProxyFacoryContract.address;
            green(`Deployed GenericProxyFactory for ${networkName} at ${genericProxyFactoryAddress}`);
            dim(`If this is not a test network, consider adding to the generic proxy factory repo constants`);
        }
        if (!genericProxyFactoryAddress) {
            throw new Error(`No GenericProxyFactory deployed for this network ()`);
        }
        cyan(`GenericProxyFactory for network ${networkName} at address ${genericProxyFactoryAddress}`);
        // grab abi and connect to contract instance
        const genericProxyFactoryContract = new ethers_1.ethers.Contract(genericProxyFactoryAddress, GenericProxyFactory_json_1.default.abi, deploySettings.signer);
        dim(`Creating Proxy...implementationAddress: ${deploySettings.implementationAddress}, initializeData: ${initializeData}`);
        const createProxyResult = yield genericProxyFactoryContract.create(deploySettings.implementationAddress, initializeData);
        dim(`Awaiting transaction confirmation...`);
        yield provider.waitForTransaction(createProxyResult.hash);
        const receipt = yield provider.getTransactionReceipt(createProxyResult.hash);
        const createdEvent = genericProxyFactoryContract.interface.parseLog(receipt.logs[0]);
        green(`Proxy for ${deploySettings.contractName} created at ${createdEvent.args.created}`);
        const bytecode = yield provider.getCode(createdEvent.args.created);
        const jsonObj = {
            address: createdEvent.args.created,
            transactionHash: receipt.transactionHash,
            receipt: receipt,
            args: deploySettings === null || deploySettings === void 0 ? void 0 : deploySettings.initializeData,
            bytecode: bytecode
        };
        const pathFileBase = `./deployments/${networkName}`;
        const pathFile = `${pathFileBase}/${deploySettings.contractName}.json`;
        fs_1.mkdirSync(pathFileBase, { recursive: true });
        fs_1.writeFileSync(pathFile, JSON.stringify(jsonObj, null, 3), { encoding: 'utf8', flag: 'w' });
        dim(`Deployments file saved at ${pathFile}`);
        return {
            newlyDeployed: true,
            address: createdEvent.args.created,
            transactionHash: createProxyResult.hash,
            receipt: receipt,
        };
    });
}
exports.factoryDeploy = factoryDeploy;
