// @idle-finance/hardhat-proposals-plugin/dist/src/ethers-contracts/factories/SimpleStorage__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleStorage__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        constant: false,
        inputs: [
            {
                name: "x",
                type: "uint256",
            },
        ],
        name: "set",
        outputs: [],
        type: "function",
    },
    {
        constant: true,
        inputs: [],
        name: "get",
        outputs: [
            {
                name: "retVal",
                type: "uint256",
            },
        ],
        type: "function",
    },
];
const _bytecode = "0x6060604052603b8060106000396000f3606060405260e060020a600035046360fe47b1811460245780636d4ce63c14602e575b005b6004356000556022565b6000546060908152602090f3";
class SimpleStorage__factory extends ethers_1.ContractFactory {
    constructor(signer) {
        super(_abi, _bytecode, signer);
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.SimpleStorage__factory = SimpleStorage__factory;
SimpleStorage__factory.bytecode = _bytecode;
SimpleStorage__factory.abi = _abi;
